File: leetcode/rust-leetcode/src/1331.rank-transform-of-an-array.rs
impl Solution {
    pub fn array_rank_transform(mut arr: Vec<i32>) -> Vec<i32> {
        if arr.len() == 0 {
            return arr;
        }
        let mut sarr = arr.to_vec();
        sarr.sort();
        let mut rank = 1;
        let mut map = std::collections::HashMap::new();
        map.insert(sarr[0], 1);
        for i in 1..sarr.len() {
            if sarr[i] != sarr[i-1] {
                rank += 1;
                map.insert(sarr[i], rank);
            }
        }
        let mut ranks = vec![1;arr.len()];
        for (i, n) in arr.iter().enumerate() {
            ranks[i] = *map.get(n).unwrap();
        }
        ranks
    }
}

File: leetcode/cpp/1331.rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        if (arr.size() == 0) {
            return {};
        }
        vector<int> arrSorted = arr;
        sort(arrSorted.begin(), arrSorted.end());
        
        unordered_map<int, int> ranks;
        ranks[arrSorted[0]] = 1;
        int rank = 2;
        for (int i = 1; i < arrSorted.size(); i++) {
            if (arrSorted[i] != arrSorted[i-1]) {
                ranks[arrSorted[i]] = rank;
                rank++;
            }
        }

        vector<int> ans(arr.size());
        for (int i=0; i < arr.size(); i++) {
            ans[i] = ranks[arr[i]];
        }
        return ans;
    }
};

